genienlp/data_utils/example.py

```diff
@@ -27,217 +27,93 @@
-            max_question_len = max(all_question_inputs_pair.length)
-            max_answer_len = max(all_answer_inputs_pair.length)
-
-        # process single examples
-        example_ids = [ex.example_id for ex in examples]
-        if override_question:
-            question_inputs = [(override_question, override_question_mask) for _ in examples]
-        else:
-            question_inputs = [(ex.question, ex.question_word_mask) for ex in examples]
-
-        if append_question_to_context_too:
-            context_inputs = [(ex.context_plus_question, ex.context_plus_question_word_mask) for ex in examples]
-        elif override_context:
-            context_inputs = [(override_context, override_context_mask) for _ in examples]
-        else:
-            context_inputs = [(ex.context, ex.context_word_mask) for ex in examples]
-            
-        answer_inputs = [(ex.answer, ex.answer_word_mask) for ex in examples]
-        
-        all_example_ids_single = example_ids
-        all_context_inputs_single = numericalizer.encode_single(context_inputs, decoder_vocab,
-                                                                max_length=max_context_len-2)
-        all_question_inputs_single = numericalizer.encode_single(question_inputs, decoder_vocab,
-                                                                 max_length=max_question_len-2)
-        all_answer_inputs_single = numericalizer.encode_single(answer_inputs, decoder_vocab,
-                                                               max_length=max_answer_len-2)
-    
-        if paired:
-            all_example_ids = all_example_ids_single + all_example_ids_pair
-            all_context_inputs = SequentialField.merge([all_context_inputs_single, all_context_inputs_pair])
-            all_question_inputs = SequentialField.merge([all_question_inputs_single, all_question_inputs_pair])
-            all_answer_inputs = SequentialField.merge([all_answer_inputs_single, all_answer_inputs_pair])
-        else:
-            all_example_ids = all_example_ids_single
-            all_context_inputs = all_context_inputs_single
-            all_question_inputs = all_question_inputs_single
-            all_answer_inputs = all_answer_inputs_single
-        return NumericalizedExamples(all_example_ids,
-                     all_context_inputs,
-                     all_question_inputs,
-                     all_answer_inputs,
-                     decoder_vocab,
-                     device, padding_function=numericalizer.pad)
+
+        for ex in examples:
+            yield NumericalizedExamples([ex.example_id],
+                                        numericalizer.encode_single(ex.context_plus_question),
+                                        numericalizer.encode_single(ex.answer))
 
     @staticmethod
-    def collate_batches(batches):
+    def collate_batches(batches : Iterable['NumericalizedExamples'], numericalizer, device):
```

Comment:
We need `''` _until_ Python 3.9, because the name `NumericalizedExamples` is not in scope at that point (due to Python's weird scoping rules for classes). All versions of Python with type annotations and the typing module support this (so IIRC 3.5+)
In Python 3.9, they introduced lazy resolution of type annotations so we would not need it.